src/main/core/scheduler/pools/unbounded.rs

```diff
@@ -0,0 +1,474 @@
+        let mut scope = WorkerScope::<'scope> {
+            pool: self,
+            _phantom: Default::default(),
+        };
+
+        let runner = TaskRunner { scope: &mut scope };
+
+        f(runner);
+    }
+}
+
+impl std::ops::Drop for UnboundedThreadPool {
+    fn drop(&mut self) {
+        self.join_internal();
+    }
+}
+
+struct WorkerScope<'scope> {
+    pool: &'scope mut UnboundedThreadPool,
+    // when we are dropped, it's like dropping the task
+    _phantom: PhantomData<Box<dyn TaskFn + 'scope>>,
+}
+
+impl<'a> std::ops::Drop for WorkerScope<'a> {
+    fn drop(&mut self) {
+        // if the task was set (if `TaskRunner::run` was called)
+        if self.pool.shared_state.task.borrow().is_some() {
+            // wait for the task to complete
+            self.pool.task_end_waiter.wait();
+
+            // clear the task
+            *self.pool.shared_state.task.borrow_mut() = None;
+
+            // generally following https://docs.rs/rayon/latest/rayon/fn.scope.html#panics
+            if self
+                .pool
+                .shared_state
+                .has_thread_panicked
+                .load(Ordering::Relaxed)
+            {
+                // we could store the thread's panic message and propagate it, but I don't think
+                // that's worth handling
+                panic!("A work thread panicked");
+            }
+        }
+    }
+}
+
+/// Allows a single task to run per pool scope.
+pub struct TaskRunner<'a, 'scope> {
+    // SAFETY: Self must be invariant over 'scope, which is why we use &mut here. See the
+    // documentation for scope() above for details.
+    scope: &'a mut WorkerScope<'scope>,
+}
+
+impl<'a, 'scope> TaskRunner<'a, 'scope> {
+    /// Run a task on the pool's threads.
+    pub fn run(self, f: impl TaskFn + 'scope) {
+        let f = Box::new(f);
+
+        // SAFETY: WorkerScope will drop this TaskFn before the end of 'scope
+        let f = unsafe {
+            std::mem::transmute::<Box<dyn TaskFn + 'scope>, Box<dyn TaskFn + 'static>>(f)
+        };
+
+        *self.scope.pool.shared_state.task.borrow_mut() = Some(f);
+
+        // start the threads
+        for semaphore in &self.scope.pool.task_start_semaphores {
+            semaphore.post();
+        }
+    }
+}
+
+fn work_loop(
+    thread_index: usize,
+    shared_state: Arc<SharedState>,
+    task_start_semaphore: LibcSemaphore,
+    mut end_counter: LatchCounter,
+) {
+    // this will poison the workpool when it's dropped
+    struct PoisonWhenDropped<'a>(&'a SharedState);
+
+    impl<'a> std::ops::Drop for PoisonWhenDropped<'a> {
+        fn drop(&mut self) {
+            // if we panicked, then inform other threads that we panicked and allow them to exit
+            // gracefully
+            self.0.has_thread_panicked.store(true, Ordering::Relaxed);
+        }
+    }
+
+    let shared_state = shared_state.as_ref();
+    let poison_when_dropped = PoisonWhenDropped(shared_state);
```

Comment:
I think this is ok, but would it be simpler / more idiomatic to use [catch_unwind](https://doc.rust-lang.org/std/panic/fn.catch_unwind.html)?

Reply:
I'm going to address this in a follow-up PR since I'll also want to change the thread-per-host scheduler at the same time.